### PR TITLE
feat: add stream support

### PR DESCRIPTION
**Pull request by zhukaihan**
### Summary

Uses the stream flag API from experiment-core to produce SDK that receives flag configs updates from server. It should replace constant polling. Currently it is disabled by default and falls back to polling if any fatal error encountered, server 501 or retry reached maximum. It retries every 15 seconds. 

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/experiments-js-server/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  no

### CODE REVIEW

File: packages/node/package.json [hunk 1]
```diff
@@ -32,6 +32,11 @@
   "dependencies": {
     "@amplitude/analytics-node": "^1.3.4",
     "@amplitude/analytics-types": "^1.3.1",
-    "@amplitude/experiment-core": "^0.7.2"
+    "@amplitude/experiment-core": "^0.7.2",
+    "eventsource": "^2.0.2"
+  },
+  "devDependencies": {
+    "@types/eventsource": "^1.1.15",
+    "@types/node": "18.8.0"
```
Comment (zhukaihan) [inline packages/node/package.json:40]: Need to pin a version of "@types/node" so compiling ts to js doesn't throw an error. I believe it's due to `eventsource` implementing `Event` which collided with node's `Event`.

File: packages/node/src/types/config.ts [hunk 2]
```diff
@@ -181,6 +219,13 @@ export const LocalEvaluationDefaults: LocalEvaluationConfig = {
   bootstrap: {},
   flagConfigPollingIntervalMillis: 30000,
   httpAgent: null,
+  getFlagConfigUpdateWithStream: false,
+  streamServerUrl: 'https://stream.lab.amplitude.com',
+  streamConnTimeoutMillis: 1000,
+  streamFlagConnTimeoutMillis: 1000,
+  streamFlagTryAttempts: 2,
+  streamFlagTryDelayMillis: 1000,
+  retryStreamFlagDelayMillis: 15000,
```
Comment (zhukaihan) [inline packages/node/src/types/config.ts:228]: Do we want to adjust these defaults? Such as reducing `streamFlagTryAttempts` to 1 for quicker start ups if stream is unavailable and increasing `retryStreamFlagDelayMillis` for longer interval between checks.

File: packages/node/package.json [hunk 3]
```diff
@@ -32,6 +32,11 @@
   "dependencies": {
     "@amplitude/analytics-node": "^1.3.4",
     "@amplitude/analytics-types": "^1.3.1",
-    "@amplitude/experiment-core": "^0.7.2"
+    "@amplitude/experiment-core": "^0.7.2",
```
Comment (zhukaihan) [inline packages/node/package.json:35]: May need a new version of `experiment-core` after that PR gets merged.

File: packages/node/package.json (hunk 1 above)
Comment (bgiori) [reply packages/node/package.json:40]: Is this going to be a problem for customers too? That could be really poor developer experience if so, and we may want to consider an alternative implementation.

Comment (zhukaihan) [reply packages/node/package.json]: Fixed by not using `@types/eventsource`. `eventsource` package is pure js. Type conflicts comes in `@types/eventsource`. We declare our own type for `eventsource` package.

File: packages/node/src/types/eventsource.d.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    dispatchEvent(evt: Event): boolean;
+    removeEventListener(
+      type: string,
+      listener: (evt: MessageEvent) => void,
+    ): void;
+    close(): void;
+  }
+
+  export default EventSource;
+}
```
Comment (zhukaihan) [inline packages/node/src/types/eventsource.d.ts:61]: Not sure if it's a good folder for this file.

File: packages/node/src/local/streamer.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+import { LocalEvaluationDefaults } from '../types/config';
+import { FlagConfigCache } from '../types/flag';
+import { ConsoleLogger } from '../util/logger';
+import { Logger } from '../util/logger';
+
+import { FlagConfigFetcher } from './fetcher';
+import { FlagConfigPoller } from './poller';
+import { SdkStreamFlagApi } from './stream-flag-api';
+
+export class FlagConfigStreamer {
```
Comment (bgiori) [inline packages/node/src/local/streamer.ts:15]: Is this supposed to `implement FlagConfigUpdater`?

File: packages/node/src/types/eventsource.d.ts
Comment (bgiori) [inline packages/node/src/types/eventsource.d.ts:1]: As of now this file is not bundled in the distributed package and types due to the file name suffix `.d.ts` which can be changed to `.ts` then the bundler will translate it to `.d.ts` in the distributed package.

You can check the `dist/src` after running `yarn`.

Comment (bgiori) [reply packages/node/src/types/eventsource.d.ts:1]: Or, if the point is to not include the types in the bundle, then this is the right approach :)

File: packages/node/src/types/config.ts [hunk 6]
```diff
@@ -148,6 +148,58 @@ export type LocalEvaluationConfig = {
...
+   */
+  streamFlagRetryDelayMillis?: number;
+
+  /**
+   * To use with streaming. The jitter to add to the delay for retry streaming
+   * after stream fatal error and fallbacked to poller.
+   * A random number between 0 and streamFlagRetryJitterMillis will be added to
+   * streamFlagRetryDelayMillis as the delay.
+   */
+  streamFlagRetryJitterMillis?: number;
```
Comment (zhukaihan) [inline packages/node/src/types/config.ts:202]: Added this jitter config.

Comment (zhukaihan) [reply packages/node/src/types/config.ts:202]: I think we can combine `streamConnTimeoutMillis` and `streamFlagConnTimeoutMillis` into one. Remove `streamFlagRetryDelayMillis` to be just default 15 seconds. Remove `streamFlagRetryJitterMillis` to just be 2 seconds jitter.

Comment (bgiori) [reply packages/node/src/types/config.ts:202]: 100% agreed. Also I think the difference between try and retry is a little unlclear, I think we only expose one of those config types. most likely just the try configs and hardcode the retry configs ourselves. 

So:
1. streamConnTimeoutMillis
2. streamFlagRetryDelayMillis

**Review by bgiori — APPROVED** [review]

**Comment by bgiori** [comment]
Just make the config changes and you're good to go

**Pull request merged**